lib/core/request.js

```diff
@@ -69,57 +69,70 @@ class Request {
+      ) {
+        this.contentLength = parseInt(val)
+        if (!Number.isFinite(this.contentLength)) {
+          throw new InvalidArgumentError('invalid content-length header')
         }
+      } else if (
+        key.length === 17 &&
+        key.toLowerCase() === 'transfer-encoding'
+      ) {
+        throw new InvalidArgumentError('invalid transfer-encoding header')
+      } else if (
+        key.length === 10 &&
+        key.toLowerCase() === 'connection'
+      ) {
+        throw new InvalidArgumentError('invalid connection header')
+      } else if (
+        key.length === 10 &&
+        key.toLowerCase() === 'keep-alive'
+      ) {
+        throw new InvalidArgumentError('invalid keep-alive header')
+      } else if (
+        key.length === 7 &&
+        key.toLowerCase() === 'upgrade'
+      ) {
+        throw new InvalidArgumentError('invalid upgrade header')
+      } else if (
+        key.length === 6 &&
+        key.toLowerCase() === 'expect'
+      ) {
+        throw new NotSupportedError('expect header not supported')
+      } else {
+        this.headers += `${key}: ${val}\r\n`
+      }
+    }
 
-        if (
-          this.host === null &&
-          key.length === 4 &&
-          key.toLowerCase() === 'host'
-        ) {
-          this.host = val
-          this.headers += `${key}: ${val}\r\n`
-        } else if (
-          this.contentLength === null &&
-          key.length === 14 &&
-          key.toLowerCase() === 'content-length'
-        ) {
-          this.contentLength = parseInt(val)
-          if (!Number.isFinite(this.contentLength)) {
-            throw new InvalidArgumentError('invalid content-length header')
-          }
-        } else if (
-          key.length === 17 &&
-          key.toLowerCase() === 'transfer-encoding'
-        ) {
-          throw new InvalidArgumentError('invalid transfer-encoding header')
-        } else if (
-          key.length === 10 &&
-          key.toLowerCase() === 'connection'
-        ) {
-          throw new InvalidArgumentError('invalid connection header')
-        } else if (
-          key.length === 10 &&
-          key.toLowerCase() === 'keep-alive'
-        ) {
-          throw new InvalidArgumentError('invalid keep-alive header')
-        } else if (
-          key.length === 7 &&
-          key.toLowerCase() === 'upgrade'
-        ) {
-          throw new InvalidArgumentError('invalid upgrade header')
-        } else if (
-          key.length === 6 &&
-          key.toLowerCase() === 'expect'
-        ) {
-          throw new NotSupportedError('expect header not supported')
-        } else {
-          this.headers += `${key}: ${val}\r\n`
+    if (headers) {
+      if (Array.isArray(headers)) {
+        if (headers.length % 2 !== 0) {
+          throw new InvalidArgumentError('headers array must be even')
+        }
+        for (let i = 0; i < headers.length; i += 2) {
+          processHeader(headers[i], headers[i + 1])
+        }
+      } else {
```

Comment:
headers && typeof headers === 'object' else throw

Reply:
this changes the API. Currently, the user doesn't have to specify any headers. Do you want me to change that?

Reply:
Oh maybe you meant if headers are not undefined and are not of type object then throw error